### PR TITLE
refactor: reduce number of ErrorKind variants

### PR DESCRIPTION
**Pull request by bartlomieju**
Partially adresses #2635

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -36,8 +36,6 @@ dirs = "2.0.2"
 dlopen = "0.1.8"
 futures = { version = "0.3.1", features = [ "compat", "io-compat" ] }
 http = "0.2.0"
-hyper = "0.13.1"
-hyper-rustls = "0.19.0"
```
Comment (ry) [inline cli/Cargo.toml]: Why? These seems unrelated?

Comment (bartlomieju) [reply cli/Cargo.toml]: Check out this comment: https://github.com/denoland/deno/issues/3607#issuecomment-573410167

File: cli/ops/permissions.rs [hunk 2]
```diff
@@ -86,7 +86,7 @@ pub fn op_request_permission(
     "env" => Ok(permissions.request_env()),
     "plugin" => Ok(permissions.request_plugin()),
     "hrtime" => Ok(permissions.request_hrtime()),
-    n => Err(type_error(format!("No such permission name: {}", n))),
+    n => Err(other_error(format!("No such permission name: {}", n))),
```
Comment (ry) [inline cli/ops/permissions.rs]: http://man7.org/linux/man-pages/man3/errno.3.html

We essentially want to map errors to something like POSIX. In this situation, I think something like EINVAL would be appropriate?

Comment (ry) [reply cli/ops/permissions.rs]: Also I think we should remove DenoError.Other - JavaScript already has generic errors TypeError, SyntaxError, etc. Shouldn't we just raise one of those if an appropriate error code isn't found?

Comment (bartlomieju) [reply cli/ops/permissions.rs]: > http://man7.org/linux/man-pages/man3/errno.3.html
> 
> We essentially want to map errors to something like POSIX. In this situation, I think something like EINVAL would be appropriate?

`ErrorKind.InvalidArgument` sounds fine for me

> Also I think we should remove DenoError.Other - JavaScript already has generic errors TypeError, SyntaxError, etc. Shouldn't we just raise one of those if an appropriate error code isn't found?

That's still gonna be needed as it's part of mechanism of forwarding errors to JS, no problem though to throw `Error` in that case instead of `DenoError`

File: cli/js/errors.ts [hunk 3]
```diff
@@ -45,39 +44,13 @@ export enum ErrorKind {
...
-  JSError = 50,
-  TypeError = 51,
-
-  /** TODO this is a DomException type, and should be moved out of here when possible */
-  DataCloneError = 52
+  UrlParse = 20,
+  Http = 21,
+  TooLarge = 22,
+  InvalidSeekMode = 23,
+  UnixError = 24,
```
Comment (piscisaureus) [inline cli/js/errors.ts]: I don't think we should fold all OS errors under 'UnixError'. It's unpractical when e.g. renaming a file fails with some generic error because you might want to handle it based on the specific reason for the failure, e.g. `rmdir` could fail because the directory was not found (so you're good) or because it's a file or because the dir is not empty.

Comment (bartlomieju) [reply cli/js/errors.ts]: Agree, it seems we've been using `UnixError` for quite some time. We should decide which of error code provided by Rust `Error::Sys` we want to support and list them collapsing the rust under `ErrorKind.SysError`

File: cli/ops/permissions.rs (hunk 2 above)
Comment (piscisaureus) [reply cli/ops/permissions.rs]: I don't think ES has a defined error class of errors that originate from the platform (e.g. ENOENT) so we might still have to define our own class for that.

For invalid URLs there's https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/URIError and I think we should definitely use that.
